### ISSUE TITLE
What should the reified type of a tear-off containing generic type parameters be?

### ISSUE THREAD

**Issue by munificent** [dart-lang/sdk#28425]
Consider:

```dart
void main() {
  List<Object> list = <int>[]; // Covariant upcast.
  var tearOff = list.add;
  tearOff = tearOff as dynamic;
}
```

The third line fails at runtime today in dev_compiler. Since it is a cast from dynamic, the cast is checked at runtime. The runtime type of the tear-off is `(int) -> void`. The statically inferred type of `tearOff` is `(Object) -> void`. Thus, the cast fails.

It seems pretty surprising that `list.add` produces a value whose runtime type is not of its static type. We are addressing this for covariant overrides by making the covariant parameters have type `Object`. Should we do the same or something similar for covariant generics?

I think so. My suggestion is that a parameter is considered "covariant" if:

* It is marked `covariant` or one of the corresponding parameters it is overriding is.
* It uses a generic type parameter in a covariant position. A "covariant" position means as the type of a parameter, the return type of a function-typed parameter, or the "flipped" case where you have an even number of levels of contravariant nesting.

If so, the parameter's reified type is `Object`.

Of course, if we ever support invariant type parameters, this would not apply to them.

**Comment by MichaelRFairhurst** [comment]
I agree that it's confusing, but my theoretical side says this is suitable. Add is a contravariant operation, and so this error seems right to me.

I do understand the debate of whether covariance/contravariance rules are that helpful given that most programmers don't understand them (and even those of us who do may take a few moments to understand why this fails). So I'll try to ignore my impression that this error is correct.

But maybe I shouldn't ignore that impression entirely...it looks to me like the problem is the statement `List<Object> list = <int>[]`. Maybe such a cast should involve creating a wrapper that checks covariance/contravariance on call (much like the `covariant` parameters do). The tearoff would then be off of the _wrapper's_ `add` method, which would have type `(Object) -> void` and the rest of the program would check out fine. And when it passes the checks, the executing code resumes from the reified context of `T=int`.

Not sure the implications of adding the wrapper in terms of the vm & more, but it would catch more misusages of generics than current.

Also, I can see a weird edge case in creating a hell of a lot of wrappers: `<int>[].add as ObjFunc as IntFunc as ObjFunc as IntFunc as ObjFunc as IntFunc`, and in fact, `<int>[].add as ObjFunc as IntFunc` must know to _not_ check the input is an `Object`, which might take more VM madness that I'm no expert on.

I have no clue how this idea on the whole could still be even slightly performant, but I'll let the vmexperts decide. Heck, maybe by adding typecheck wrappers, the JIT can produce better code in the real methods!

**Comment by munificent** [comment]
> it looks to me like the problem is the statement List<Object> list = <int>[]

Yes, the fact that we allow covariance on all generics even when it's obviously unsound is the core of the issue.

But, assuming we don't change that for now, we have to decide how the tear-off of one of these incorrectly covariant methods should behave.

> Maybe such a cast should involve creating a wrapper that checks covariance/contravariance on call (much like the covariant parameters do).

I think we don't want to wrap on assignment because it ends up causing lots of problems around nested wrappers and broken identity. Users should be able to safely assume that hoisting a subexpression to a variable still gives you the same object.

> The tearoff would then be off of the wrapper's add method, which would have type (Object) -> void and the rest of the program would check out fine.

Right. But we can also do that by doing the "wrapping" at the point that the tear-off is taken, which is what I suggest here. Because taking a tear-off is already "producing" some kind of object, I think it's reasonable if it produces one whose reified type is looser and then it does the argument check in the body (which is what we need to keep things sound).

> Also, I can see a weird edge case in creating a hell of a lot of wrappers: <int>[].add as ObjFunc as IntFunc as ObjFunc as IntFunc as ObjFunc as IntFunc, and in fact, <int>[].add as ObjFunc as IntFunc must know to not check the input is an Object, which might take more VM madness that I'm no expert on.

Exactly right. Not just the VM either—we need to be able to compile to sane JS.

**Comment by MichaelRFairhurst** [comment]
Alternatively, the "we'll let you cast covariantly and then watch that you use it correctly" thing might just not be being used _enough_.

Maybe assigning `(T) -> X` to `(Z) -> X` where `Z :> T` shouldn't produce an error. Maybe it should wait until the function is called with a `Z` (or subtype) where `T` was expected to report an error. This matches the philosophy of waiting to see if an int was added to throw.

Always reifying as Object also makes the object contravariant, too. That seems wrong to me. Does that apply to `covariant` already in other places?

**Comment by munificent** [comment]
> Always reifying as Object also makes the object contravariant, too. That seems wrong to me.

I didn't follow this. Can you give me an example?

**Comment by MichaelRFairhurst** [comment]
I'm probably just not understanding the topic since I don't know the nitty gritty parts of dart's type system, but my understanding is this...

If condition `A` means that all` X<T>` should be treated as `X<Object>`, then all `X` are `X<Object>`, meaning `X<int>` and `X<Object>` are the same type. And the same type is always substitutable for itself, meaning you can use `X<int>` in place of `X<Object>` and vice versa, making it both covariant and contravariant.

In fact, worse than that, two types with no relationship would be "the same," `X<int>` and `X<String>`. Which is actually a good sign to me that I'm misunderstanding the implications of reifying `T` as `Object` in this case.

**Comment by eernstg** [comment]
It's a transitional issue.

What is happening is that strong mode replaces the subtype rule for function types: current Dart uses covariance for parameter types (which is unsound but consistent with the generic covariance), and strong mode uses the traditional (sound) rule where parameter types are contravariant.

As a consequence, the dynamic value of a tear-off of a method in a generic class using a type argument in a contravariant position (e.g., as the type of a parameter) is capable of yielding a value which does not have a subtype of the statically known type. This violates a property that Dart otherwise maintains (in anything but production mode): an expression evaluation yields a value which is a subtype of the statically known type (or null).

So we need to make further adjustments (in addition to the new function type subtype rule) to make the pieces fit together again.

One plausible way to restore the expression type soundness property is to adjust the reified type of the tear-off such that we _do_ get a subtype of the statically known type, and a particularly robust and simple way to achieve this is to let the reified parameter type of such "covariant" parameters be `Object`. This is what Bob describes above. Another meaningful choice is to use the least upper bound of all the types declared for that parameter (taking all overridden methods where that parameter occurs into account). The former is more flexible (the tear-off is allowed to be the value of a variable with a larger set of types) and the latter is more precise (disallowing some types where invocations are unlikely to succeed, thus "catching the invocation failure earlier"). But both approaches maintain heap soundness, because the actual arguments are subject to dynamic type checks whenever there is an invocation of a function with a covariant parameter.

Consider Bob's example again:
```dart
void main() {
  List<Object> list = <int>[]; // Covariant upcast.
  var tearOff = list.add;
  tearOff = tearOff as dynamic; // Downcast to the inferred type of `tearOff`.
}
```
In the current situation, it shouldn't be the cast from `dynamic` to `Object -> void` that fails, it should be the initialization of `tearOff` (and then of course the cast is never reached). With the adjusted reified type it will all succeed. But if we add an invocation like `tearOff('Hello!')` at the end, it would be statically safe and still it would throw a `CastError` at runtime. That's the way it has always been in Dart, and it's a deliberate soundness/convenience trade-off.

Note that there is no need for any wrappers: A function with a covariant parameter will have a reified type where that parameter is specified to have a sufficiently general type to make the function type as a whole sufficiently special to ensure that we maintain the expression soundness property. An actual argument for such a covariant parameter will be subject to a dynamic type check, so the body of the function will be able to safely use the parameter with the declared type.

**Comment by eernstg** [comment]
Let me see if I can clarify this a bit.

> I'm probably just not understanding the topic since I don't know the nitty gritty parts of dart's 
> type system, but my understanding is this...
>
> If condition A means that all `X<T>` should be treated as `X<Object>`, then all X are `X<Object>`,
> meaning `X<int>` and `X<Object>` are the same type. And the same type is always substitutable
> for itself, meaning you can use `X<int>` in place of `X<Object>` and vice versa, making it both
> covariant and contravariant.

I couldn't see what condition A refers to, but type arguments of generic classes are reified at runtime and they do matter during static analysis, so it is certainly not the case that `X<int>` and `X<Object>` are identical types.

For instance, `X<int> x = new X<Object>();` is a downcast at compile time and a run-time error, whereas `X<Object> x = new X<int>();` is an upcast statically and will run without errors at run time (unless the constructor body throws, of course). The point is that generic classes in Dart are covariant in every type parameter, which means that `X<T2>` is a subtype of `X<T1>` whenever `T2` is a subtype of `T1`.

> In fact, worse than that, two types with no relationship would be "the same," X<int> and
> X<String>. Which is actually a good sign to me that I'm misunderstanding the implications of
> reifying T as Object in this case.

We never considered reifying type arguments of generic classes as anything other than their actual value.

The reified type we are discussing here is much more subtle: For an instance of a generic class (say `List<num> xs = <int>[];`), when a method is torn off (`var f = xs.add`), the closure which is the value of evaluating `xs.add` has a type which is a function type. Statically, it would be based on the statically known information about the receiver (so `xs` is known to be a `List<num>` and hence its `add` method has type `num -> void`). Dynamically, the naive approach would be to say that it should have type `int -> void`, because that's how it is declared in the class `List` when the type argument `E` is `int`. However, this would mean that the dynamic type of the expression `xs.add` is a **supertype** of the static type, and that contradicts the expression soundness property that I mentioned. So we can't use that approach.

The main approach on the table is to reify that parameter type as `Object`.

That's the only case where reification as `Object` is considered here: In a reified type for a torn-off method, in the part of that type which specifies the type of a covariant parameter. In particular, type arguments of generic classes are treated the same way as always, so they are distinct when you'd expect it, but covariant.

**Comment by MichaelRFairhurst** [comment]
(in regards to condition A, that's part my-mind-wandering, but I think I needed to end my sentence with "...covariant and contravariant _under condition A_". Apologies).

I think I understand the actual suggestion better now, thanks!

So essentially, upon tearoff of a method which has a covariable* type (such as a generic type parameter) in a contravariant placement (such as a method parameter), reify that type inside the tearoff as `Object` so that it acts covariant.

*made up the word covariable to include both covariant in theory and covariant in dart :)

```
  List<num> list = <int>[];
  var add =  list.add; // has type (Object) -> void, now both statically and at runtime
  // equivalent to var add = ((x) => (Object y) => y is int ? x.add(y) : throw ...)(list)
```

Hopefully this time I've followed that right so far. I'm not sure if my "equivalent to" section is actually equivalent, or if it uses an unnecessary wrapper. `x.add(y)` might already insert a typecheck. If its marked covariant I know it will include a typecheck.

Wait...are tearoffs instance members? like `void myMethod() {...}` implicitly adds some hidden symbol `Function myMethodTearoff = () { myMethod(); }`? If so, I think `Object` makes sense. Though it certainly has some quirks.

```
  List<num> list = <int>[];
  List<String> strlist = <String>[];
  var add = strlist.add; // of type Object -> void
  add = list.add; // no error, oven though List<num>.add and List<String>.add should have no relationship
```

If tearoffs are created on the fly, you reify as the static type, in this case `num`.

```
  List<num> list = <int>[];
  var add1 = <String>[].add; // static type is String -> void, so no change
  var add2 = list.add; // static type is num -> void, so type reification result is different
  add = list.add; // error: num -> void not a subtype of String -> void
```

Basically, once the tearoff is created it can't go any higher than the static type, since the tearoff itself is a function which has good typing rules.

Though I can see this having quirks, too.

```
  List<int> list = <int>[];
  var add1 = list.add; // int -> void
  List<Object> upcast = list;
  var add2 = upcast.add; // Object -> void
  print(identical(add1, add2)); // how can they be identical and reified differently? But how could they not be identical?
  add2 = add1; // Implicit upcast, passes static analysis but throws at runtime.
```

Oh wait, turns out tearoffs already don't report `identical`. So I guess they aren't members like I had theorized?

**Comment by eernstg** [comment]
> .. reify that type inside the tearoff as `Object` ..

Yes, to make it completely explicit: The parameter type is as declared when the function is invoked (so we check that we actually get an `int` argument with the `<int>[].add`), but the reified type of the tear-off is a function type where the covariant parameter types have been replaced by `Object` (so `<int>[].add` has the reified type `(Object) -> void`).

> .. `has type (Object) -> void, now both statically and at runtime` ..

No, the static type is actually `(num) -> void`. We use the available static information (here: `list` has type `List<num>`, or a subtype thereof). This means that we statically prevent assigning such a tear-off to some variables where the assignment would actually succeed (e.g., `((f(Object)) => f('xyz')) list.add` is rejected), but in those cases the _dynamic_ check is unlikely to succeed (e.g., the invocation passing a string to `f` will fail). If you insist, you can just upcast the receiver (`(<int>[] as List<Object>).add` has static type `(Object) -> void`).

The 'equivalent to' expression can be simpler: `(Object x) { int y = x; list.add(y); }`, where the initialization of `y` may throw because it's a downcast.

In the next example, `add = list.add` will be a static type error: The inferred type of `add` is `(String) -> void` and the static type of `list.add` is `(num) -> void`: unrelated. You may still cheat it (one step) by explicitly declaring `add` to be an `(Object) -> void`; statically it's a downcast to initialize that with `strlist.add`, but it succeeds at runtime; same for `add = list.add`. For invocations (that's step two) you can still have failures with some arguments due to the dynamic argument check, with both tear-offs.

Similar adjustments are needed for the following examples.

Finally, tear-offs do not have to be identical even when they are tear-offs of the same method (that is: same method name and same receiver), but they do have to be equal. In your `identical(add1, add2)`, `add1` and `add2` are indeed based on the same method name and receiver, so they must be equal and can be identical. Both have reified type `(Object) -> void` and a dynamic check for `int`.

**Comment by MichaelRFairhurst** [comment]
OK, I should plainly ask...
example1:
```
  List<num> list = <int>[];
  var x = list.add;
```

What is being suggested? That `x` in example1 has a dynamic reified type of `Object`, or `num`? And if the former, why? Its confusing when using the upper bound of `Object`, which is the same as not having an upper bound.

I'll say strategy 1 results in `Object` and strategy 2 results in `num`.

> No, the static type is actually (num) -> void

Meant to say, dynamic type of `Object -> void`, static type of `num -> void`. Assuming Bob is suggesting strategy 1.

> In the next example, add = list.add will be a static type error:

Ah, yes, of course. My mistake there. But if we go with strategy 1, then this will succeed at runtime if you forget to run the typechecker. Which is less bad but still weird. I know it would throw when used later, but I see no reason to delay that error until after assignment. Under strategy 2 it will be both a static and runtime error.

> tear-offs do not have to be identical....but they do have to be equal

My point with the identicality is based on conjecture. I like strategy 2, but I might be missing a good reason to consider strategy 1. I tried to think of such reasons, and thought that maybe if tearoffs were implemented like so:

```
class X<T> {
  void y(T z) { ... }
  // behind the scenes adds
  final Function yTearoff = (T z) { y(z); };
  ...
}

var x = new X<T>;
x.y; // equivalent to x.yTearoff;
```

Then I could see two potential problems. One is that a single final value could not satisfy all static contexts, meaning a new implementation would be required to reify each tearoff at call time rather than at construction time.

Secondly, in this hypothetical implementation, `identical(x.y, x.y)` would be true. Naively, a new implementation to achieve strategy 2 would change that. Maintaining that identicality while using strategy 2 might be difficult and/or hurt performance elsewhere. Its a good thing for strategy 2 that there doesn't seem to be any such behavior to maintain.

**Comment by leafpetersen** [comment]
@eernstg This is resolved right - I think you wrote up an informal spec for this?  Do we still need language tests and some follow up to get it implemented?

**Comment by eernstg** [comment]
Right. In strong mode, the reified function types discussed in this issue are specified in [the informal spec](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/covariant-from-class.md), and they have `Object` as the reified type for each parameter
